File: typescript/tests/test_a2a_agent_executor.ts
/** Tests for A2A AgentExecutor implementation. */
/* eslint-disable @typescript-eslint/unbound-method */

import { describe, it, expect, vi, beforeEach } from 'vitest';
import { TimestepAgentExecutor } from '../timestep/a2a/agent_executor.js';
import {
  extractTextFromParts,
  a2AToOpenAI as convertA2AMessagesToOpenAI,
} from '../timestep/a2a/message_converter.js';
import type { RequestContext, ExecutionEventBus } from '@a2a-js/sdk/server';
import type { Message, Task } from '@a2a-js/sdk';
import { runAgent } from '../timestep/core/agent.js';

// Mock the agent module
vi.mock('../timestep/core/agent.js', () => ({
  runAgent: vi.fn(),
}));

describe('A2A AgentExecutor Helper Functions', () => {
  it('should extract text from parts', () => {
    /** Test extracting text from A2A message parts. */
    const parts = [
      { kind: 'text', text: 'Hello' },
      { kind: 'text', text: ' World' },
      { kind: 'other', data: 'ignored' },
    ];
    const result = extractTextFromParts(parts);
    expect(result).toBe('Hello World');
  });

  it('should convert A2A messages to OpenAI format', () => {
    /** Test converting A2A messages to OpenAI format. */
    const a2aMessages = [
      {
        kind: 'message',
        role: 'user',
        messageId: 'msg-1',
        parts: [{ kind: 'text', text: 'Hello' }],
      } as Message,
      {
        kind: 'message',
        role: 'agent',
        messageId: 'msg-2',
        parts: [{ kind: 'text', text: 'Hi there' }],
      } as Message,
    ];
    const result = convertA2AMessagesToOpenAI(a2aMessages);
    
    expect(result.length).toBe(2);
    expect(result[0].role).toBe('user');
    expect(result[0].content).toBe('Hello');
    expect(result[1].role).toBe('assistant');
    expect(result[1].content).toBe('Hi there');
  });

  it('should convert A2A messages with tool calls', () => {
    /** Test converting A2A messages with tool calls. */
    const a2aMessages = [
      {
        kind: 'message',
        role: 'user',
        messageId: 'msg-1',
        parts: [{ kind: 'text', text: "What's the weather?" }],
      } as Message,
      {
        kind: 'message',
        role: 'agent',
        messageId: 'msg-2',
        parts: [{ kind: 'text', text: '' }],
        toolCalls: [{ id: 'call_1', function: { name: 'get_weather', arguments: '{"city": "Oakland"}' } }],
      } as Message & { toolCalls: Array<{ id: string; function: { name: string; arguments: string } }> },
      {
        kind: 'message',
        role: 'tool',
        messageId: 'msg-3',
        parts: [{ kind: 'text', text: 'Sunny' }],
        toolCallId: 'call_1',
      } as unknown as Message & { toolCallId: string },
    ];
    const result = convertA2AMessagesToOpenAI(a2aMessages);
    
    expect(result.length).toBe(3);
    expect(result[0].role).toBe('user');
    expect(result[0].content).toBe("What's the weather?");
    expect(result[1].role).toBe('assistant');
    expect('tool_calls' in result[1]).toBe(true);
    expect(result[2].role).toBe('tool');
    expect((result[2] as { tool_call_id: string }).tool_call_id).toBe('call_1');
  });
});

describe('A2A AgentExecutor Tests', () => {
  let executor: TimestepAgentExecutor;
  let mockEventBus: ExecutionEventBus;
  let mockContext: RequestContext;

  beforeEach(() => {
    executor = new TimestepAgentExecutor();
    mockEventBus = {
      publish: vi.fn(),
    } as unknown as ExecutionEventBus;
    
    mockContext = {
      userMessage: {
        kind: 'message',
        role: 'user',
        messageId: 'test-msg-1',
        parts: [{ kind: 'text', text: 'Hello' }],
        contextId: 'test-context-id',
      },
      task: undefined,
    } as RequestContext;
  });

  it('should initialize with default tools', () => {
    /** Test TimestepAgentExecutor initialization. */
    expect(executor).toBeDefined();
    // Tools are private, but we can verify by checking execution
  });

  it('should create a task when none exists', async () => {
    /** Test that executor creates a task when none exists. */
    vi.mocked(runAgent).mockResolvedValue('Test response');

    await executor.execute(mockContext, mockEventBus);

    // Verify task was created (first publish should be a task)
    expect(mockEventBus.publish).toHaveBeenCalled();
    const firstCall = vi.mocked(mockEventBus.publish).mock.calls[0];
    expect(firstCall).toBeDefined();
    const firstEvent = firstCall?.[0] as { kind?: string; status?: { state?: string } };
    if (firstEvent?.kind === 'task' && firstEvent?.status) {
      expect(firstEvent.status.state).toBe('submitted');
    }
  });

  it('should use existing task when provided', async () => {
    /** Test executor with existing task. */
    const existingTask: Task = {
      kind: 'task',
      id: 'test-task-id',
      contextId: 'test-context-id',
      status: {
        state: 'submitted',
        timestamp: new Date().toISOString(),
      },
      history: [],
    };

    Object.assign(mockContext, { task: existingTask });
    vi.mocked(runAgent).mockResolvedValue('Test response');

    await executor.execute(mockContext, mockEventBus);

    // Verify runAgent was called
    expect(runAgent).toHaveBeenCalled();
    // Verify events were published
    expect(mockEventBus.publish).toHaveBeenCalled();
  });

  it('should publish working status update', async () => {
    /** Test that executor publishes working status. */
    vi.mocked(runAgent).mockResolvedValue('Test response');

    await executor.execute(mockContext, mockEventBus);

    // Find working status update
    const calls = vi.mocked(mockEventBus.publish).mock.calls;
    const workingUpdate = calls.find(
      (call) => {
        const event = call[0] as { kind?: string; status?: { state?: string } };
        return event?.kind === 'status-update' && event?.status?.state === 'working';
      }
    );
    expect(workingUpdate).toBeDefined();
  });

  it('should publish completed status on success', async () => {
    /** Test that executor publishes completed status. */
    vi.mocked(runAgent).mockResolvedValue('Test response');

    await executor.execute(mockContext, mockEventBus);

    // Find completed status update
    const calls = vi.mocked(mockEventBus.publish).mock.calls;
    const completedUpdate = calls.find(
      (call) => {
        const event = call[0] as { kind?: string; status?: { state?: string }; final?: boolean };
        return event?.kind === 'status-update' && 
               event?.status?.state === 'completed' &&
               event?.final === true;
      }
    );
    expect(completedUpdate).toBeDefined();
  });

  it('should publish failed status on error', async () => {
    /** Test executor error handling. */
    vi.mocked(runAgent).mockRejectedValue(new Error('Test error'));

    // The executor catches errors and publishes failed status, it doesn't throw
    await executor.execute(mockContext, mockEventBus);

    // Find failed status update
    const calls = vi.mocked(mockEventBus.publish).mock.calls;
    const failedUpdate = calls.find(
      (call) => {
        const event = call[0] as { kind?: string; status?: { state?: string }; final?: boolean };
        return event?.kind === 'status-update' && 
               event?.status?.state === 'failed' &&
               event?.final === true;
      }
    );
    expect(failedUpdate).toBeDefined();
  });

  it('should cancel task', async () => {
    /** Test executor cancellation. */
    const taskId = 'test-task-id';
    
    await executor.cancelTask(taskId, mockEventBus);

    // Verify cancellation event was published
    expect(mockEventBus.publish).toHaveBeenCalled();
    const call = vi.mocked(mockEventBus.publish).mock.calls[0];
    const event = call?.[0] as { kind?: string; status?: { state?: string }; final?: boolean };
    expect(event?.kind).toBe('status-update');
    expect(event?.status?.state).toBe('canceled');
    expect(event?.final).toBe(true);
  });

  it('should handle tool approval requirement', async () => {
    /** Test tool approval handling. */
    // Mock runAgent to trigger tool approval via event emitter
    vi.mocked(runAgent).mockImplementation(async (
      _messages: unknown,
      _tools: unknown,
      _model: unknown,
      _apiKey: unknown,
      eventEmitter: import('../timestep/core/agent_events.js').AgentEventEmitter | undefined
    ) => {
      if (eventEmitter) {
        // Simulate tool call
        const toolCall = {
          id: 'call_1',
          type: 'function' as const,
          function: {
            name: 'get_weather',
            arguments: '{"city": "Oakland"}',
          },
        };
        // Emit tool-approval-required event - this will create a Future and wait for it
        const future = new Promise<boolean>((resolve) => {
          // Emit with the correct format that the handler expects
          eventEmitter.emit('tool-approval-required', {
            toolCall,
            resolve,
          });
          // Give it a moment to set up the resolver and publish the event
          setTimeout(() => {
            // Manually resolve the approval by finding the resolver and calling it
            const pendingApprovals = executor['_pendingApprovals'] as Map<string, { toolCall: unknown; resolve: (approved: boolean) => void }> | undefined;
            if (pendingApprovals && pendingApprovals.size > 0) {
              const approval = Array.from(pendingApprovals.values())[0];
              approval.resolve(true); // Auto-approve for test
            }
          }, 100);
        });
        
        // Wait a bit to let the status be published
        await new Promise(resolve => setTimeout(resolve, 100));
        
        // Wait for approval to complete
        await future;
      }
      return 'Test response';
    });

    // Execute - this will trigger the approval requirement
    const executePromise = executor.execute(mockContext, mockEventBus);
    
    // Wait for the input-required status to be published
    await new Promise(resolve => setTimeout(resolve, 100));

    // Verify input-required status was published
    const calls = vi.mocked(mockEventBus.publish).mock.calls;
    const inputRequiredUpdate = calls.find(
      (call) => {
        const event = call[0] as { kind?: string; status?: { state?: string } };
        return event?.kind === 'status-update' && 
               event?.status?.state === 'input-required';
      }
    );
    expect(inputRequiredUpdate).toBeDefined();
    
    // Now simulate approval by sending an approval message to resolve the pending approval
    // Get task ID from the published events
    const taskEvent = calls.find(
      (call) => {
        const event = call[0] as { kind?: string; taskId?: string };
        return event?.kind === 'task' && event?.taskId;
      }
    );
    const taskId = taskEvent 
      ? (taskEvent[0] as { taskId?: string })?.taskId || 'test-task-id'
      : 'test-task-id';
    
    const approvalContext = {
      ...mockContext,
      userMessage: {
        ...mockContext.userMessage,
        parts: [{ kind: 'text', text: 'approve' }],
        contextId: 'test-context-id',
        taskId: taskId,
      },
    };
    
    // Execute with approval message - this should resolve the pending approval and continue
    await executor.execute(approvalContext, mockEventBus);
    
    // Wait for the original execution to complete
    await executePromise;
  }, 10000); // Increase timeout for this test
});

File: python/tests/test_a2a_agent_executor.py
"""Tests for A2A AgentExecutor implementation."""

import pytest
from unittest.mock import AsyncMock, MagicMock, patch
from a2a.server.agent_execution import RequestContext
from a2a.server.events import EventQueue
from a2a.types import Message, Task, TaskState

from timestep.a2a.agent_executor import TimestepAgentExecutor
from timestep.a2a.message_converter import (
    a2a_to_openai as _convert_a2a_messages_to_openai,
    extract_text_from_parts as _extract_text_from_parts,
)


def test_extract_text_from_parts():
    """Test extracting text from A2A message parts."""
    parts = [
        {"kind": "text", "text": "Hello"},
        {"kind": "text", "text": " World"},
        {"kind": "other", "data": "ignored"},
    ]
    result = _extract_text_from_parts(parts)
    assert result == "Hello World"


def test_convert_a2a_messages_to_openai():
    """Test converting A2A messages to OpenAI format."""
    a2a_messages = [
        {
            "role": "user",
            "parts": [{"kind": "text", "text": "Hello"}],
        },
        {
            "role": "agent",
            "parts": [{"kind": "text", "text": "Hi there"}],
        },
    ]
    result = _convert_a2a_messages_to_openai(a2a_messages)
    
    assert len(result) == 2
    assert result[0]["role"] == "user"
    assert result[0]["content"] == "Hello"
    assert result[1]["role"] == "assistant"
    assert result[1]["content"] == "Hi there"


def test_convert_a2a_messages_with_tool():
    """Test converting A2A messages with tool calls."""
    a2a_messages = [
        {
            "role": "user",
            "parts": [{"kind": "text", "text": "What's the weather?"}],
        },
        {
            "role": "agent",
            "parts": [{"kind": "text", "text": ""}],
            "tool_calls": [{"id": "call_1", "function": {"name": "get_weather", "arguments": '{"city": "Oakland"}'}}],
        },
        {
            "role": "tool",
            "tool_call_id": "call_1",
            "parts": [{"kind": "text", "text": "Sunny"}],
        },
    ]
    result = _convert_a2a_messages_to_openai(a2a_messages)
    
    assert len(result) == 3
    assert result[0]["role"] == "user"
    assert result[0]["content"] == "What's the weather?"
    assert result[1]["role"] == "assistant"
    assert "tool_calls" in result[1]
    assert result[2]["role"] == "tool"
    assert result[2]["tool_call_id"] == "call_1"


@pytest.mark.asyncio
async def test_agent_executor_initialization():
    """Test TimestepAgentExecutor initialization."""
    executor = TimestepAgentExecutor()
    assert executor.tools is not None
    assert len(executor.tools) == 2  # GetWeather and WebSearch
    assert executor.model == "gpt-4.1"


@pytest.mark.asyncio
async def test_agent_executor_execute_creates_task():
    """Test that executor creates a task when none exists."""
    executor = TimestepAgentExecutor()
    
    # Create proper mock message object
    from a2a.types import TextPart
    mock_message = Message(
        kind="message",
        role="user",
        messageId="test-msg-1",
        parts=[TextPart(kind="text", text="Hello")],
        contextId="test-context-id",
    )
    
    mock_context = MagicMock(spec=RequestContext)
    mock_context.current_task = None
    mock_context.message = mock_message
    mock_context.get_user_input = MagicMock(return_value="Hello")
    mock_context.context_id = "test-context-id"
    
    mock_event_queue = MagicMock(spec=EventQueue)
    mock_event_queue.enqueue_event = AsyncMock()
    
    # Mock run_agent to return immediately
    with patch("timestep.a2a.agent_executor.run_agent", new_callable=AsyncMock) as mock_run:
        mock_run.return_value = "Test response"
        
        await executor.execute(mock_context, mock_event_queue)
        
        # Verify task was created
        assert mock_event_queue.enqueue_event.call_count >= 2  # Task creation + working + completed
        # First call should be task creation
        first_call = mock_event_queue.enqueue_event.call_args_list[0]
        assert first_call is not None


@pytest.mark.asyncio
async def test_agent_executor_execute_with_existing_task():
    """Test executor with existing task."""
    executor = TimestepAgentExecutor()
    
    # Create mock task
    mock_task = MagicMock(spec=Task)
    mock_task.id = "test-task-id"
    mock_task.context_id = "test-context-id"
    mock_task.history = []
    
    mock_message = MagicMock()
    mock_message.parts = [MagicMock(kind="text", text="Hello")]
    mock_message.role = "user"
    
    mock_context = MagicMock(spec=RequestContext)
    mock_context.current_task = mock_task
    mock_context.message = mock_message
    mock_context.get_user_input = MagicMock(return_value="Hello")
    
    mock_event_queue = MagicMock(spec=EventQueue)
    mock_event_queue.enqueue_event = AsyncMock()
    
    with patch("timestep.a2a.agent_executor.run_agent", new_callable=AsyncMock) as mock_run:
        mock_run.return_value = "Test response"
        
        await executor.execute(mock_context, mock_event_queue)
        
        # Verify run_agent was called
        mock_run.assert_called_once()
        # Verify events were published
        assert mock_event_queue.enqueue_event.call_count >= 2


@pytest.mark.asyncio
async def test_agent_executor_cancel():
    """Test executor cancellation."""
    executor = TimestepAgentExecutor()
    
    mock_task = MagicMock(spec=Task)
    mock_task.id = "test-task-id"
    mock_task.context_id = "test-context-id"
    
    mock_context = MagicMock(spec=RequestContext)
    mock_context.current_task = mock_task
    
    mock_event_queue = MagicMock(spec=EventQueue)
    mock_event_queue.enqueue_event = AsyncMock()
    
    await executor.cancel(mock_context, mock_event_queue)
    
    # Verify cancellation event was published
    mock_event_queue.enqueue_event.assert_called_once()
    call_args = mock_event_queue.enqueue_event.call_args[0][0]
    assert call_args.status.state == TaskState.canceled
    assert call_args.final is True


@pytest.mark.asyncio
async def test_agent_executor_error_handling():
    """Test executor error handling."""
    executor = TimestepAgentExecutor()
    
    from a2a.types import TextPart
    mock_message = Message(
        kind="message",
        role="user",
        messageId="test-msg-1",
        parts=[TextPart(kind="text", text="Hello")],
        contextId="test-context-id",
    )
    
    mock_context = MagicMock(spec=RequestContext)
    mock_context.current_task = None
    mock_context.message = mock_message
    mock_context.get_user_input = MagicMock(return_value="Hello")
    mock_context.context_id = "test-context-id"
    
    mock_event_queue = MagicMock(spec=EventQueue)
    mock_event_queue.enqueue_event = AsyncMock()
    
    # Mock run_agent to raise an error
    with patch("timestep.a2a.agent_executor.run_agent", new_callable=AsyncMock) as mock_run:
        mock_run.side_effect = Exception("Test error")
        
        # Execute should not raise - it catches and publishes error event
        await executor.execute(mock_context, mock_event_queue)
        
        # Verify error event was published
        call_args_list = mock_event_queue.enqueue_event.call_args_list
        # Find the error event (should have state "failed")
        error_event = None
        for call in call_args_list:
            event = call[0][0]
            if hasattr(event, "status") and event.status.state == TaskState.failed:
                error_event = event
                break
        
        assert error_event is not None
        assert error_event.final is True


@pytest.mark.asyncio
async def test_agent_executor_publishes_working_status():
    """Test that executor publishes working status update."""
    executor = TimestepAgentExecutor()
    
    from a2a.types import TextPart
    mock_message = Message(
        kind="message",
        role="user",
        messageId="test-msg-1",
        parts=[TextPart(kind="text", text="Hello")],
        contextId="test-context-id",
    )
    
    mock_context = MagicMock(spec=RequestContext)
    mock_context.current_task = None
    mock_context.message = mock_message
    mock_context.get_user_input = MagicMock(return_value="Hello")
    mock_context.context_id = "test-context-id"
    
    mock_event_queue = MagicMock(spec=EventQueue)
    mock_event_queue.enqueue_event = AsyncMock()
    
    with patch("timestep.a2a.agent_executor.run_agent", new_callable=AsyncMock) as mock_run:
        mock_run.return_value = "Test response"
        
        await executor.execute(mock_context, mock_event_queue)
        
        # Find working status update
        call_args_list = mock_event_queue.enqueue_event.call_args_list
        working_update = None
        for call in call_args_list:
            event = call[0][0]
            if hasattr(event, "status") and event.status.state == TaskState.working:
                working_update = event
                break
        
        assert working_update is not None
        assert working_update.final is False


@pytest.mark.asyncio
async def test_agent_executor_publishes_completed_status():
    """Test that executor publishes completed status on success."""
    executor = TimestepAgentExecutor()
    
    from a2a.types import TextPart
    mock_message = Message(
        kind="message",
        role="user",
        messageId="test-msg-1",
        parts=[TextPart(kind="text", text="Hello")],
        contextId="test-context-id",
    )
    
    mock_context = MagicMock(spec=RequestContext)
    mock_context.current_task = None
    mock_context.message = mock_message
    mock_context.get_user_input = MagicMock(return_value="Hello")
    mock_context.context_id = "test-context-id"
    
    mock_event_queue = MagicMock(spec=EventQueue)
    mock_event_queue.enqueue_event = AsyncMock()
    
    with patch("timestep.a2a.agent_executor.run_agent", new_callable=AsyncMock) as mock_run:
        mock_run.return_value = "Test response"
        
        await executor.execute(mock_context, mock_event_queue)
        
        # Find completed status update
        call_args_list = mock_event_queue.enqueue_event.call_args_list
        completed_update = None
        for call in call_args_list:
            event = call[0][0]
            if hasattr(event, "status") and event.status.state == TaskState.completed:
                completed_update = event
                break
        
        assert completed_update is not None
        assert completed_update.final is True


@pytest.mark.asyncio
async def test_agent_executor_tool_approval_requirement():
    """Test tool approval requirement handling."""
    executor = TimestepAgentExecutor()
    
    from a2a.types import TextPart
    mock_message = Message(
        kind="message",
        role="user",
        messageId="test-msg-1",
        parts=[TextPart(kind="text", text="What's the weather?")],
        contextId="test-context-id",
    )
    
    mock_context = MagicMock(spec=RequestContext)
    mock_context.current_task = None
    mock_context.message = mock_message
    mock_context.get_user_input = MagicMock(return_value="What's the weather?")
    mock_context.context_id = "test-context-id"
    
    mock_event_queue = MagicMock(spec=EventQueue)
    mock_event_queue.enqueue_event = AsyncMock()
    
    import asyncio
    from timestep.core.agent_events import AgentEventEmitter
    
    # Mock run_agent to trigger tool approval via event emitter
    async def mock_run_agent_with_approval(*args, **kwargs):
        event_emitter = kwargs.get("event_emitter")
        if event_emitter:
            tool_call = {
                "id": "test-tool-call-id",
                "type": "function",
                "function": {
                    "name": "get_weather",
                    "arguments": '{"city": "Oakland"}',
                },
            }
            # Emit tool-approval-required event - this will create a Future and wait for it
            future = asyncio.Future()
            # Emit the event - this will trigger the handler which is async
            await event_emitter.emit_async("tool-approval-required", {
                "tool_call": tool_call,
                "resolve": lambda approved: future.set_result(approved),
            })
            # Give it a moment to set up the resolver and publish the event
            await asyncio.sleep(0.1)
            # Manually resolve the approval by finding the resolver and calling it
            if executor._pending_approvals:
                # Get the first (and only) approval
                approval = next(iter(executor._pending_approvals.values()))
                approval["resolve"](True)  # Auto-approve for test
            # Wait for approval to complete (with timeout)
            try:
                await asyncio.wait_for(future, timeout=1.0)
            except asyncio.TimeoutError:
                # If timeout, just return
                pass
        return "Test response"
    
    with patch("timestep.a2a.agent_executor.run_agent", new_callable=AsyncMock) as mock_run:
        mock_run.side_effect = mock_run_agent_with_approval
        
        await executor.execute(mock_context, mock_event_queue)
        
        # Verify input-required status was published
        call_args_list = mock_event_queue.enqueue_event.call_args_list
        input_required_update = None
        for call in call_args_list:
            event = call[0][0]
            if hasattr(event, "status") and event.status.state == TaskState.input_required:
                input_required_update = event
                break
        
        assert input_required_update is not None
        assert input_required_update.final is False  # Should be False to keep stream open

